packages/api/lib/executions.js

```diff
@@ -0,0 +1,55 @@
+const {
+  getKnexClient,
+  executionArnsFromGranuleIdsAndWorkflowNames,
+} = require('@cumulus/db');
+
+const { RecordDoesNotExist } = require('@cumulus/errors');
+
+/**
+ *  Finds and returns alternative executionArn related to the input granuleId.
+ *  Used to override the default (latest) executionArn when reingesting granules.
+ *  The decision tree is simple.
+ *  1. If a user inputs an executionArn we return it.
+ *  2. If not and no workflowName is specified, we return undefined so that the
+ *  granule's original execution is retained during reingest.
+ *  3. if not and a workflowName is input, we search the database for all
+ *  executions that match the granuleId and workflowName and return the most
+ *  recent.
+ *
+ * @param {Object} params - function parameters
+ * @param {string} params.granuleId - granuleId
+ * @param {string|undefined} [params.executionArn] - execution arn to use for reingest
+ * @param {string|undefined} [params.workflowName] - workflow name to use for reingest
+ * @param {function|undefined} [params.dbFunction] - database function for
+ *     testing. Defaults to executionArnsFromGranuleIdsAndWorkflowNames.
+ * @returns {Promise<string>|Promise<undefined>} - executionArn used in a
+ *             granule reingest call to determine correct workflow to run or
+ *             undefined.
+ */
+const chooseTargetExecution = async ({
+  granuleId,
+  executionArn = undefined,
+  workflowName = undefined,
+  dbFunction = executionArnsFromGranuleIdsAndWorkflowNames,
+}) => {
+  // if a user specified an executionArn, use that always
+  if (executionArn !== undefined) return executionArn;
```

Comment:
It will fail if the executionArn is bad.  I feel like this is / should be user beware, since if we did that we'd be validating every execution in the database before re-running.  In the 99% case this call should come from the dashboard.  In a bulk-reingest this could be a very large number of executions.

I think you need to convince me.  I was trying to think of the things that will be the most common and not change default behavior and keep overhead down.

Reply:
Nope! That totally makes sense to me. I think you're completely right that validating the execution would be a ton of overhead in the case of bulk-reingest. Hadn't thought of that. 

It sounds like, if the request is coming from the dashboard, the odds of receiving an invalid executionArn are pretty negligible. If that's the case and/or is pretty obvious when this is the issue, I don't think you need additional validation here. I buy the "user beware" approach.

Reply:
sold.